File: backend/app/schemas/rule.py
"""规则 schemas（D05 §5）。"""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from app.schemas.common import RuleAction, RuleStatus


class RuleCreate(BaseModel):
    """POST /rules 请求体（D05 §5.2）。"""

    name: str
    description: str | None = None
    dsl: str
    severity: str = "WARN"
    action: str = "REVIEW"
    valid_from: datetime | None = None
    valid_to: datetime | None = None
    scope: dict[str, Any] = Field(default_factory=dict)


class RuleUpdate(BaseModel):
    """PUT /rules/{rule_id} 请求体（D05 §5.4）。"""

    name: str | None = None
    description: str | None = None
    dsl: str | None = None
    severity: str | None = None
    action: str | None = None
    valid_from: datetime | None = None
    valid_to: datetime | None = None
    scope: dict[str, Any] | None = None


class RuleOut(BaseModel):
    """规则详情。"""

    id: str
    rule_id: str
    name: str
    description: str | None = None
    dsl: str
    severity: str
    action: str
    status: RuleStatus
    version: str
    valid_from: datetime | None = None
    valid_to: datetime | None = None
    scope: dict[str, Any] = Field(default_factory=dict)
    hit_count_24h: int = 0
    false_positive_rate: float = 0.0
    created_at: datetime
    updated_at: datetime


class RuleVersionCreate(BaseModel):
    """POST /rules/{rule_id}/versions 请求体（D05 §5.6）。"""

    dsl: str
    change_summary: str | None = None
    severity: str | None = None
    action: str | None = None


class RuleVersionOut(BaseModel):
    """规则版本。"""

    id: str
    rule_id: str
    version: str
    dsl: str
    status: RuleStatus
    canary_percent: int = 0
    created_by: str
    created_at: datetime
    promoted_at: datetime | None = None


class RulePromoteRequest(BaseModel):
    """POST /rules/{rule_id}/promote 请求体（D05 §5.7）。"""

    from_status: RuleStatus
    to_status: RuleStatus
    canary_percentage: int | None = None
    approver_id: str
    observation_hours: int | None = None
    rollback_thresholds: dict[str, Any] | None = None


class RuleRollbackRequest(BaseModel):
    """POST /rules/{rule_id}/rollback 请求体（D05 §5.8）。"""

    target_version: int | None = None
    reason: str
    approver_id: str


__all__ = [
    "RuleCreate",
    "RuleOut",
    "RulePromoteRequest",
    "RuleRollbackRequest",
    "RuleUpdate",
    "RuleVersionCreate",
    "RuleVersionOut",
]

File: frontend/src/types/rule.ts
/**
 * 规则引擎类型
 * 对齐 D05 §5（规则引擎接口）与 baseline §3.4（规则状态/动作枚举）
 */
import type { RuleStatus, RuleAction, RuleSeverity, Channel } from './enum'

/** 规则列表查询参数（D05 §5.1） */
export interface RuleQuery {
  status?: RuleStatus
  action?: RuleAction
  channel?: Channel
  severity?: RuleSeverity
}

/** 规则列表项（D05 §5.1） */
export interface RuleListItem {
  rule_id: string
  name: string
  description?: string
  dsl: string
  severity: RuleSeverity
  action: RuleAction
  status: RuleStatus
  version: number
  hit_count_24h: number
  false_positive_rate: number
  created_at: string
  updated_at: string
}

/** 规则版本历史（D05 §5.3） */
export interface RuleVersion {
  version: number
  dsl: string
  status: RuleStatus
  change_summary?: string
  created_at: string
  created_by: string
}

/** 规则详情（D05 §5.3） */
export interface RuleDetail extends RuleListItem {
  valid_from?: string
  valid_to?: string | null
  scope?: { channels: Channel[] }
  published_at?: string
  published_by?: string
  versions: RuleVersion[]
}

/** 创建/更新规则请求（D05 §5.2 / §5.4） */
export interface UpsertRuleRequest {
  name: string
  description?: string
  dsl: string
  severity: RuleSeverity
  action: RuleAction
  valid_from?: string
  valid_to?: string | null
  scope?: { channels: Channel[] }
}

/** 创建新版本请求（D05 §5.6） */
export interface CreateRuleVersionRequest {
  dsl: string
  change_summary?: string
  severity?: RuleSeverity
  action?: RuleAction
}

/** 灰度推进请求（D05 §5.7） */
export interface PromoteRuleRequest {
  from_status: RuleStatus.DRAFT | RuleStatus.CANARY
  to_status: RuleStatus.CANARY | RuleStatus.ACTIVE
  canary_percentage?: number
  approver_id: string
  observation_hours?: number
  rollback_thresholds?: {
    false_positive_rate?: number
    precision_drop?: number
  }
}

/** 回滚请求（D05 §5.8） */
export interface RollbackRuleRequest {
  target_version?: number
  reason: string
  approver_id: string
}

/** DSL 校验请求（D05 §5.9） */
export interface ValidateRuleDslRequest {
  dsl: string
  sample_transactions?: string[]
}

/** DSL 校验响应 */
export interface ValidateRuleDslResult {
  valid: boolean
  syntax_errors: string[]
  sample_hits: { external_tx_id: string; matched: boolean; evaluated_at_ms: number }[]
}

/** 规则历史命中（D05 §5.11） */
export interface RuleHitRecord {
  hit_id: string
  rule_id: string
  external_tx_id: string
  decision: string
  risk_score: number
  occurred_at: string
}
